### PR TITLE
Add sanity check for number of threads and block size

### LINKED ISSUE #60194: Logical error (Too large size) on high max_block_size
Fuzz: https://s3.amazonaws.com/clickhouse-test-reports/60191/9bf3d6ae2927f5d1b4ffdccde2da93a73518c133/ast_fuzzer__asan_.html

```sql
    SELECT zero + 1 AS x
    FROM system.zeros
    SETTINGS max_block_size = 9223372036854775806, max_rows_to_read = 20, read_overflow_mode = 'break'
```

```
(gdb) bt
#0  __pthread_kill_implementation (threadid=<optimized out>, signo=signo@entry=6, no_tid=no_tid@entry=0) at pthread_kill.c:44
#1  0x00007ffff7e35393 in __pthread_kill_internal (signo=6, threadid=<optimized out>) at pthread_kill.c:78
#2  0x00007ffff7de46c8 in __GI_raise (sig=sig@entry=6) at ../sysdeps/posix/raise.c:26
#3  0x00007ffff7dcc4b8 in __GI_abort () at abort.c:79
#4  0x0000000014df299f in DB::abortOnFailedAssertion (description=...) at /mnt/ch/ClickHouse/src/Common/Exception.cpp:43
#5  0x0000000014df2a75 in DB::handle_error_code (msg=..., code=49, remote=false, trace=...) at /mnt/ch/ClickHouse/src/Common/Exception.cpp:59
#6  0x0000000014df307f in DB::Exception::Exception (this=0x7ffff2f17680, msg_masked=..., code=49, remote_=false) at /mnt/ch/ClickHouse/src/Common/Exception.cpp:100
#7  0x000000000adcf137 in DB::Exception::Exception (this=0x7ffff2f17680, msg=..., code=49, remote_=false) at /mnt/ch/ClickHouse/src/Common/Exception.h:90
#8  0x000000000afadd84 in DB::Exception::Exception<unsigned long&> (this=0x7ffff2f17680, code=49, fmt=..., args=@0x7ffffffeb2c0: 9223372036854775933) at /mnt/ch/ClickHouse/src/Common/Exception.h:109
#9  0x0000000014d85024 in (anonymous namespace)::checkSize (size=9223372036854775933) at /mnt/ch/ClickHouse/src/Common/Allocator.cpp:101
#10 0x0000000014d8680a in Allocator<false, false>::alloc (this=0x7ffff505bae0, size=9223372036854775933, alignment=0) at /mnt/ch/ClickHouse/src/Common/Allocator.cpp:113
#11 0x000000000ade3c65 in DB::PODArrayBase<1ul, 4096ul, Allocator<false, false>, 63ul, 64ul>::alloc<>(unsigned long) (this=0x7ffff505bae0, bytes=9223372036854775933) at /mnt/ch/ClickHouse/src/Common/PODArray.h:115
#12 0x000000000ade3bb9 in DB::PODArrayBase<1ul, 4096ul, Allocator<false, false>, 63ul, 64ul>::realloc<>(unsigned long) (this=0x7ffff505bae0, bytes=9223372036854775933) at /mnt/ch/ClickHouse/src/Common/PODArray.h:140
#13 0x0000000014e34db5 in DB::PODArrayBase<1ul, 4096ul, Allocator<false, false>, 63ul, 64ul>::reserve<>(unsigned long) (this=0x7ffff505bae0, n=9223372036854775806) at /mnt/ch/ClickHouse/src/Common/PODArray.h:218
#14 DB::PODArray<char8_t, 4096ul, Allocator<false, false>, 63ul, 64ul>::resize_fill (this=0x7ffff505bae0, n=9223372036854775806) at /mnt/ch/ClickHouse/src/Common/PODArray.h:402
#15 0x000000001b66e45c in DB::(anonymous namespace)::ZerosSource::createColumn (size=9223372036854775806) at /mnt/ch/ClickHouse/src/Storages/System/StorageSystemZeros.cpp:76
#16 0x000000001b66df62 in DB::(anonymous namespace)::ZerosSource::ZerosSource (this=0x7ffff31031d8, block_size=9223372036854775806, limit_=0, state_=...) at /mnt/ch/ClickHouse/src/Storages/System/StorageSystemZeros.cpp:32
#17 0x000000001b66ddff in std::__1::construct_at[abi:v15000]<DB::(anonymous namespace)::ZerosSource, unsigned long&, unsigned long, std::__1::shared_ptr<DB::(anonymous namespace)::ZerosState>&, DB::(anonymous namespace)::ZerosSource*>(DB::(anonymous namespace)::ZerosSource*, unsigned long&, unsigned long&&, std::__1::shared_ptr<DB::(anonymous namespace)::ZerosState>&) (__location=0x7ffff31031d8, __args=..., __args=..., __args=...)
    at /mnt/ch/ClickHouse/contrib/llvm-project/libcxx/include/__memory/construct_at.h:35
#18 0x000000001b66db31 in std::__1::allocator_traits<std::__1::allocator<DB::(anonymous namespace)::ZerosSource> >::construct[abi:v15000]<DB::(anonymous namespace)::ZerosSource, unsigned long&, unsigned long, std::__1::shared_ptr<DB::(anonymous namespace)::ZerosState>&, void, void>(std::__1::allocator<DB::(anonymous namespace)::ZerosSource>&, DB::(anonymous namespace)::ZerosSource*, unsigned long&, unsigned long&&, std::__1::shared_ptr<DB::(anonymous namespace)::ZerosState>&) (__p=0x7ffff31031d8, 
    __args=..., __args=..., __args=...) at /mnt/ch/ClickHouse/contrib/llvm-project/libcxx/include/__memory/allocator_traits.h:298
#19 0x000000001b66d7e4 in std::__1::__shared_ptr_emplace<DB::(anonymous namespace)::ZerosSource, std::__1::allocator<DB::(anonymous namespace)::ZerosSource> >::__shared_ptr_emplace[abi:v15000]<unsigned long&, unsigned long, std::__1::shared_ptr<DB::(anonymous namespace)::ZerosState>&>(unsigned long&, unsigned long&&, std::__1::shared_ptr<DB::(anonymous namespace)::ZerosState>&) (this=0x7ffff31031c0, __a=..., __args=..., __args=..., __args=...)
    at /mnt/ch/ClickHouse/contrib/llvm-project/libcxx/include/__memory/shared_ptr.h:292
#20 0x000000001b66d5b7 in std::__1::allocate_shared[abi:v15000]<DB::(anonymous namespace)::ZerosSource, std::__1::allocator<DB::(anonymous namespace)::ZerosSource>, unsigned long&, unsigned long, std::__1::shared_ptr<DB::(anonymous namespace)::ZerosState>&, void>(std::__1::allocator<DB::(anonymous namespace)::ZerosSource> const&, unsigned long&, unsigned long&&, std::__1::shared_ptr<DB::(anonymous namespace)::ZerosState>&) (__a=..., __args=..., __args=..., __args=...)
    at /mnt/ch/ClickHouse/contrib/llvm-project/libcxx/include/__memory/shared_ptr.h:953
#21 0x000000001b66c9a3 in std::__1::make_shared[abi:v15000]<DB::(anonymous namespace)::ZerosSource, unsigned long&, unsigned long, std::__1::shared_ptr<DB::(anonymous namespace)::ZerosState>&, void>(unsigned long&, unsigned long&&, std::__1::shared_ptr<DB::(anonymous namespace)::ZerosState>&) (__args=..., __args=..., __args=...) at /mnt/ch/ClickHouse/contrib/llvm-project/libcxx/include/__memory/shared_ptr.h:962
#22 0x000000001b66c639 in DB::StorageSystemZeros::read (this=0x7ffff6923e18, column_names=..., storage_snapshot=..., max_block_size=9223372036854775806, num_streams=1) at /mnt/ch/ClickHouse/src/Storages/System/StorageSystemZeros.cpp:124
#23 0x000000001f4d3e2f in DB::IStorage::read (this=0x7ffff6923e18, query_plan=..., column_names=..., storage_snapshot=..., query_info=..., context=..., processed_stage=DB::QueryProcessingStage::FetchColumns, max_block_size=9223372036854775806, 
    num_streams=16) at /mnt/ch/ClickHouse/src/Storages/IStorage.cpp:145
```

We should have some sanity check around max_block_size since it has a direct relationship with memory usage.

### LINKED ISSUE #60280: Huge value of max_threads breaks merge tree select
```
create table a (i int) engine MergeTree order by i as select * from numbers(10);
select * from a settings max_threads = 9223371761976868863;  -- Ok
select * from a settings max_threads = 9223371761976868864;  -- Fail
```

```
Code: 1001. DB::Exception: Received from localhost:9000. DB::Exception: std::length_error: vector. Stack trace:

0. ./contrib/llvm-project/libcxx/include/exception:141: std::logic_error::logic_error(char const*) @ 0x00000000158c2ff4 in programs/clickhouse
1. ./contrib/llvm-project/libcxx/include/stdexcept:154: std::length_error::length_error[abi:v15000](char const*) @ 0x00000000065c2309 in programs/clickhouse
2. ./contrib/llvm-project/libcxx/include/stdexcept:257: std::__throw_length_error[abi:v15000](char const*) @ 0x00000000065c22bf in programs/clickhouse
3. ? @ 0x0000000011f4060b in programs/clickhouse
4. ./contrib/llvm-project/libcxx/include/__memory/allocator.h:108: std::vector<DB::MergeTreeReadPool::ThreadTask, std::allocator<DB::MergeTreeReadPool::ThreadTask>>::__append(unsigned long) @ 0x0000000011f405ea in programs/clickhouse
5. ./contrib/llvm-project/libcxx/include/vector:0: DB::MergeTreeReadPool::fillPerThreadInfo(unsigned long, unsigned long) @ 0x0000000011f3c95c in programs/clickhouse
6. ./src/Storages/MergeTree/MergeTreeReadPool.cpp:88: DB::MergeTreeReadPool::MergeTreeReadPool(DB::RangesInDataParts&&, std::shared_ptr<DB::StorageSnapshot> const&, std::shared_ptr<DB::PrewhereInfo> const&, DB::ExpressionActionsSettings const&, DB::MergeTreeReaderSettings const&, std::vector<String, std::allocator<String>> const&, std::vector<String, std::allocator<String>> const&, DB::MergeTreeReadPoolBase::PoolSettings const&, std::shared_ptr<DB::Context const> const&) @ 0x0000000011f3c806 in programs/clickhouse
```

### PR DESCRIPTION
**Pull request by Algunenano**
### Changelog category (leave one):
- Backward Incompatible Change

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
- Add sanity check for number of threads and block sizes

### Documentation entry for user-facing changes

https://s3.amazonaws.com/clickhouse-test-reports/58738/6329b6ae459519bc0471d3ce0179ed2bd09dacca/ast_fuzzer__ubsan_.html

Closes https://github.com/ClickHouse/ClickHouse/issues/60194
Closes https://github.com/ClickHouse/ClickHouse/issues/60280

### CODE REVIEW

**Comment by robot-ch-test-poll1** [comment]
*This is an automated comment for commit da9ec13d5b3040b45331460cacc9380059f539f1 with description of existing statuses. It's updated for the latest CI running*

[⏳ Click here](https://s3.amazonaws.com/clickhouse-test-reports/60138/da9ec13d5b3040b45331460cacc9380059f539f1/ci_running.html) to open a full report in a separate page

<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>CI running</td><td>A meta-check that indicates the running CI. Normally, it's in <b>success</b> or <b>pending</b> state. The failed status indicates some problems with the PR</td><td><a href="https://s3.amazonaws.com/clickhouse-test-reports/60138/da9ec13d5b3040b45331460cacc9380059f539f1/ci_running.html">⏳ pending</a></td></tr>
<tbody>
</table>

**Review by alexey-milovidov — APPROVED** [review]

**Comment by alexey-milovidov** [comment]
Also, check the place where `length_error` happens. If there is an overflow, we should also check for it.
I assume, it can gracefully return error - either something like "cannot allocate a thread" or "memory limit exceeded" - let's check it on multiple levels.

**Comment by Algunenano** [comment]
Seeing https://github.com/ClickHouse/ClickHouse/issues/60194 maybe I should move sanity checks to a common place like SettingsConstraints::check or have an always on setting profile called `Sanity`. I think I prefer having `SettingsConstraints::sanityCheck` though

File: tests/queries/0_stateless/01730_distributed_group_by_no_merge_order_by_long.sql [hunk 1]
```diff
@@ -4,15 +4,15 @@ drop table if exists data_01730;
 
 -- does not use 127.1 due to prefer_localhost_replica
 
-select * from remote('127.{2..11}', view(select * from numbers(1e6))) group by number order by number limit 20 settings distributed_group_by_no_merge=0, max_memory_usage='100Mi'; -- { serverError 241 }
+select * from remote('127.{2..11}', view(select * from numbers(1e6))) group by number order by number limit 20 settings distributed_group_by_no_merge=0, max_memory_usage='200Mi'; -- { serverError 241 }
```
Comment (Algunenano) [inline tests/queries/0_stateless/01730_distributed_group_by_no_merge_order_by_long.sql:7]: This is wrong. It was me testing to see where the memory raised to adapt the second query (which is still tricky to adjust).

**Comment by clickgapai** [comment]
Hi @Algunenano @alexey-milovidov — while reviewing this PR I found the following:

- **Coverage gap:** `doSettingsSanityCheckClamp` `max_threads` clamping branch untested in stateless SQL — https://github.com/ClickHouse/ClickHouse/pull/103877

Could you review the test PR and add the `can be tested` label if it looks good?
Happy to discuss — close anything that's wrong or already addressed.

**Pull request merged**